### PR TITLE
Optimize Docker image for model serving

### LINKED ISSUE #2426: [FR] Reduce Docker Image Size for Predict
### Motivation
Docker Image produced by MLFlow for predicting the models is supposed to be the most portable format but unfortunately, the docker image produced by MLFlow is of huge size. We tried with a simple model file with 1kb size (a simple regression model) but the resultant docker image is 3.3 GB.  We are trying to deploy these docker images (about 60) in Kubernetes but since K8s nodes will have limited resources, docker images of this size make mlflow a non-viable option. 
### Proposed Changes
Please let us know how the docker image size for predicting the model can be reduced.

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/10954?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10954/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10954
```

</p>
</details>

### Related Issues/PRs

Resolve #2426, #5927

#### Background
Docker image for model serving was not optimized and its large size has been pain for users (#2426, #5927). For example, few MB of scikit-learn model becomes > 3.7GB image after `build-image` command. Also the long building time (> 5 minutes) has been another pain.

After internal design discussion, we decided to introduce three optimizations:
1. Remove Java and dependencies if the model flavor doesn't require them.
2. Use Python base image instead of Ubuntu if possible.
3. Remove unnecessary conda/virtualenv layer if using Python image.
This PR implements those changes, as well as comprehensive integration tests with (almost) all flavors.

#### Implementation Notes
Some cases we cannot apply optimization, e.g. model_uri is not specified. To simplify the logic, I implemented those conditions as "all-or-nothing", namely, apply 1+2+3 all or do nothing. 
The optimizations apply when all the following conditions are met:
1. `model_uri` is specified.
2. User doesn't enable `--install-java` flag. (new CLI param)
3. The model flavor is not one of those require Java (e.g. `spark`, `mleap`)
4. Python version can be determined from the model metadata.
If any of this isn't met, we fallback to the original Ubuntu image with virtualenv and java. Tho majority of usage should meet these conditions so can benefit from the optimization.

#### Testing
For better safety, I've added integration tests covers (almost) all flavors in `tests/pyfunc/test_docker_flavors.py`. These tests are a bit time consuming (~20 mins) so I mark them to be skipped in the CI.

#### Impact
The impact depends on the actual size of necessary dependencies and model files, but significant for small models. For example, an image for small scikit-learn model becomes 0.99 GB (vs 3.7GB original; 73% reduction). Build time also decreases to 67 secs (vs 430 secs original; 84% reduction). The margin could be bigger for thin/small models like OpenAI, Langchain, while smaller for large models like Pytorch, Transformers.

#### For Reviewers
While the number of changed file is a bit large (28), most of them doesn't require deep review, such as test Dockerfiles, small tweak to fixture naming (to avoid conflict). Basically core logic changes reside following three iles:
* `mlflow/models/container/__init__.py`
* `mlflow/models/docker_utils.py`
* `mlflow/pyfunc/backend.py`

#### Can we reduce the image size further?
Speaking of the scikit-learn image, the majority of the size is used by MLflow itself and its dependencies (0.79 GB out of 0.99GB total), while many modules and dependencies are not required for model serving.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Screenshot 2024-02-01 at 13 40 42

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

I will update deployment doc with the change accordingly (user-facing change shouldn't be too big).

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Optimize Docker image for model serving. This reduces image size ~80% and build time ~85% depending on the model size and flavor. Note that this includes default behavior change for Java installation; the Docker image doesn't include Java by default, for all model flavors except Spark, Mleap, JohnSnowLabs, H2O. Please specify `--install-java` flag to enforce installation.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [x] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [x] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -89,6 +89,7 @@ extend-exclude = [
 "examples/*" = ["T20"]
 "docs/broken_links.py" = ["T20"]
 "mlflow/*" = ["PT018"]
+"tests/pyfunc/docker/test_docker_flavors.py" = ["F401"] # For pytest fixture imports
```
Comment (harupy) [inline pyproject.toml:92]:
```suggestion
```

We can add `noqa: F401` on lines violating this rule.

File: tests/pyfunc/docker/__init__.py
Comment (harupy) [inline tests/pyfunc/docker/__init__.py:1]: do we need this file?

File: tests/pyfunc/docker/conftest.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        pass
+
+
+@lru_cache(maxsize=1)
+def get_released_mlflow_version():
+    url = "https://pypi.org/pypi/mlflow/json"
+    response = requests.get(url)
+    response.raise_for_status()
+    data = response.json()
+    versions = [Version(v) for v in data["releases"].keys() if not v.endswith(("a", "b", "rc"))]
```
Comment (harupy) [inline tests/pyfunc/docker/conftest.py:38]:
```suggestion
    versions = [v for v in map(Version, data["releases"]) if not (v.is_devreleae or v.is_prerelease)]
```

File: tests/pyfunc/docker/test_docker.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+from tests.pyfunc.docker.conftest import RESOURCE_DIR, get_released_mlflow_version
+
+
+def assert_dockerfiles_equal(actual_dockerfile_path: Path, expected_dockerfile_path: Path):
+    actual_dockerfile = actual_dockerfile_path.read_text().replace(
+        VERSION, get_released_mlflow_version()
+    )
+    expected_dockerfile = (
+        expected_dockerfile_path.read_text()
+        .replace("$MLFLOW_VERSION", get_released_mlflow_version())
```
Comment (harupy) [inline tests/pyfunc/docker/test_docker.py:30]:
```suggestion
        .replace("${{ MLFLOW_VERSION }}", get_released_mlflow_version())
```

Can we use `${{ ... }}` as a placeholder? `$...` looks like it's an environment variable.

File: tests/pyfunc/docker/conftest.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    try:
+        # Get all containers using the test image
+        containers = docker_client.containers.list(filters={"ancestor": TEST_IMAGE_NAME})
+        for container in containers:
+            container.remove(force=True)
+
+        # Clean up the image
+        docker_client.images.remove(TEST_IMAGE_NAME, force=True)
+    except Exception:
+        pass
```
Comment (harupy) [inline tests/pyfunc/docker/conftest.py:29]: Do we need this try-catch?

File: mlflow/models/container/__init__.py [hunk 5]
```diff
@@ -136,6 +113,57 @@ def _install_pyfunc_deps(
...
+        return []
+    env_conf = conf[mlflow.pyfunc.ENV]
+
+    if env_manager == em.LOCAL:
+        # Install pip dependencies directly into the local environment
+        python_env_config_path = os.path.join(model_path, env_conf[em.VIRTUALENV])
+        python_env = _PythonEnv.from_yaml(python_env_config_path)
+        deps = " ".join(python_env.build_dependencies + python_env.dependencies)
+        deps = deps.replace("requirements.txt", os.path.join(model_path, "requirements.txt"))
+        if Popen(["bash", "-c", f"python -m pip install {deps}"]).wait() != 0:
```
Comment (harupy) [inline mlflow/models/container/__init__.py:139]:
```suggestion
        if Popen(["python", "-c", ...]).wait() != 0:
```

we can remove `bash -c` unless we need to run multiple commands joined with `&&`.

File: tests/pyfunc/docker/test_docker.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+_TEST_PARAMS = [
+    Param(expected_dockerfile="Dockerfile_default"),
+    Param(install_mlflow=True, expected_dockerfile="Dockerfile_install_mlflow"),
+    Param(enable_mlserver=True, expected_dockerfile="Dockerfile_enable_mlserver"),
+    Param(mlflow_home=".", expected_dockerfile="Dockerfile_with_mlflow_home"),
+    Param(specify_model_uri=False, expected_dockerfile="Dockerfile_no_model_uri"),
+]
+
+
+@pytest.mark.parametrize("params", _TEST_PARAMS)
```
Comment (harupy) [inline tests/pyfunc/docker/test_docker.py:81]:
```suggestion
@pytest.mark.parametrize("params", [
    Param(expected_dockerfile="Dockerfile_default"),
    Param(install_mlflow=True, expected_dockerfile="Dockerfile_install_mlflow"),
    Param(enable_mlserver=True, expected_dockerfile="Dockerfile_enable_mlserver"),
    Param(mlflow_home=".", expected_dockerfile="Dockerfile_with_mlflow_home"),
    Param(specify_model_uri=False, expected_dockerfile="Dockerfile_no_model_uri"),
])
```

File: tests/diviner/test_diviner_model_export.py [hunk 7]
```diff
@@ -79,6 +79,17 @@ def diviner_custom_env(tmp_path):
...
+@pytest.fixture
+def diviner_groups(diviner_data):
+    groups = []
+    for i in [0, -1]:
+        key_entries = []
+        for value in diviner_data.df[diviner_data.key_columns].iloc[[i]].to_dict().values():
+            key_entries.append(list(value.values())[0])
+        groups.append(tuple(key_entries))
+    return groups
+
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:91]: Can we move this change to a separate PR?

File: pyproject.toml (hunk 1 above)
Comment (B-Step62) [reply pyproject.toml:92]: yea I did that at first but was a bit tedious to repeat it for many lines....

File: tests/diviner/test_diviner_model_export.py (hunk 7 above)
Comment (B-Step62) [reply tests/diviner/test_diviner_model_export.py:91]: Not directly necessary but this is for convenience to use the fixture in the docker flavor tests.

File: tests/pyfunc/docker/__init__.py
Comment (B-Step62) [reply tests/pyfunc/docker/__init__.py:1]: nope, will remove!

File: tests/pyfunc/docker/conftest.py (hunk 4 above)
Comment (B-Step62) [reply tests/pyfunc/docker/conftest.py:29]: Yes, If test failed before creating docker image, this clean up throws exception and hides what actually failed in the test.

File: tests/pyfunc/docker/test_docker.py (hunk 3 above)
Comment (B-Step62) [reply tests/pyfunc/docker/test_docker.py:30]: Makes sense! Will replace them.

File: mlflow/models/docker_utils.py [hunk 8]
```diff
@@ -30,106 +22,112 @@
...
     """
-    mlflow_home = os.path.abspath(mlflow_home) if mlflow_home else None
+    if base_image.startswith("python:"):
+        setup_python_venv_steps = (
+            "RUN apt-get -y update && apt-get install -y --no-install-recommends nginx"
+        )
+        setup_java_steps = ""
+        install_mlflow_steps = _pip_mlflow_install_step(output_dir, mlflow_home)
+
+    elif base_image == "ubuntu:20.04":
```
Comment (BenWilson2) [inline mlflow/models/docker_utils.py:89]: Do we need this to be hard-coded? Can this just check against the constant `UBUNTU_BASE_IMAGE` ?

File: mlflow/models/python_api.py [hunk 9]
```diff
@@ -19,6 +19,7 @@ def build_docker(
     name="mlflow-pyfunc",
     env_manager=_EnvManager.VIRTUALENV,
     mlflow_home=None,
+    install_java=False,
```
Comment (BenWilson2) [inline mlflow/models/python_api.py:22]: Can we add a note block within the pydoc string to explain that this argument needs to be explicitly enabled to support container builds for SparkML and other JVM-dependent model flavors?

File: tests/pyfunc/docker/test_docker_flavors.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+
+    # Run a container
+    port = get_safe_port()
+    docker_client.containers.run(
+        image=TEST_IMAGE_NAME,
+        ports={8080: port},
+        detach=True,
+    )
+
+    # Wait until the container to start
```
Comment (BenWilson2) [inline tests/pyfunc/docker/test_docker_flavors.py:106]:
```suggestion
    # Wait for the container to start
```

**Comment by BenWilson2** [comment]
Overall this looks great! Can we:
- Verify that this builds in a windows environment 
- Create a followup ticket to add full usage and customization documentation around the introduced changes (Java being removed from the containers, now creating an opt-in experience for inclusion) into the relevant sub section tutorials in https://www.mlflow.org/docs/latest/deployment/index.html 
- Setup something similar to cross version testing that is run weekly on the mlflow-automation repo so that we have recurring CI validation of the container build process (I 100% agree that this should not be part of the PR CI process, but we should have some non-manual testing mechanism)

File: mlflow/models/python_api.py (hunk 9 above)
Comment (B-Step62) [reply mlflow/models/python_api.py:22]: Nope this doesn't need to be enabled for those flavors (we have a static list of flavors and always install java for them). This is mainly for custom pyfunc use case where we don't know users' model need java or not.

**Comment by B-Step62** [comment]
Thanks for the review, @BenWilson2!

> Verify that this builds in a windows environment

Sure, do you know what is the easiest option to test this? remote desktop?

> Create a followup ticket to add full usage and customization documentation around the introduced changes (Java being removed from the containers, now creating an opt-in experience for inclusion) into the relevant sub section tutorials in https://www.mlflow.org/docs/latest/deployment/index.html

Yup, but Java is not totally removed, they are still installed for flavors like spark, mleap. The flag will be used for custom pyfunc mode, but will add documentation for it anyway.


> Setup something similar to cross version testing that is run weekly on the mlflow-automation repo so that we have recurring CI validation of the container build process (I 100% agree that this should not be part of the PR CI process, but we should have some non-manual testing mechanism

Totally makes sense, will do this as a part of follow-up.

File: mlflow/models/python_api.py (hunk 9 above)
Comment (BenWilson2) [reply mlflow/models/python_api.py:22]: Right, that's what I was getting at :) Custom PythonModel implementation where users are wrapping a SparkML implementation for deployment as a container on AML / Sagemaker. There are a number of organizations that do this and this is effectively a breaking change for their workflows. We'll just want to make it clear in the docs and the release notes that this functionality is changing.

File: mlflow/models/container/__init__.py (hunk 5 above)
Comment (B-Step62) [reply mlflow/models/container/__init__.py:139]: Supposedly if we remove `bash -c`, we need to pass each requirement as a separate element in the list, (cuz each item passed to `Popen` is quoted together). It's doable but keeping `bash -c` is simpler imo.

**Comment by BenWilson2** [comment]
Can we see if it's possible to test really quick with https://azure.microsoft.com/en-us/products/virtual-desktop (This doesn't have to be a CI job; it's just safer to do a one-time check to see if there is any odd behavior when trying to build this PR's implementation on Windows - just to be safe :) )

**Review by BenWilson2 — APPROVED** [review]
LGTM once we verify that there aren't any surprises for Windows users :)

File: mlflow/models/python_api.py (hunk 9 above)
Comment (B-Step62) [reply mlflow/models/python_api.py:22]: Ah makes sense, thanks for the explanation! Will add more docstring in this PR, and update docs accordingly in a follow-up task.

File: mlflow/models/container/__init__.py (hunk 5 above)
Comment (harupy) [reply mlflow/models/container/__init__.py:139]: Got it, let's keep `bash -c`.

File: tests/pyfunc/docker/conftest.py (hunk 4 above)
Comment (harupy) [reply tests/pyfunc/docker/conftest.py:29]: Is it possible to check if the image exists?

Comment (B-Step62) [reply tests/pyfunc/docker/conftest.py:29]: We can't directly check with image_name (`docker_client.image.get(xxx)` does similar but raises an exception if not exists). We can use `docker_client.image.list()` and search it, but imo it's simpler to just remove inside try-catch as long as it does no harm.

File: tests/pyfunc/docker/test_docker_flavors.py [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+            # Send health check
+            response = requests.get(url=f"http://localhost:{port}/ping")
+            if response.status_code == 200:
+                success = True
+                break
+        except Exception:
+            time.sleep(5)
+
+    if not success:
+        raise TimeoutError(f"Container did not start in {timeout} seconds")
```
Comment (harupy) [inline tests/pyfunc/docker/test_docker_flavors.py:121]: Can we simplify this? For example:

```python
for _ in range(30):
    try:
        response = requests.get(url=f"http://localhost:{port}/ping")
        if response.ok:
            break
    except requests.exceptions.ConnectionError:
        time.sleep(5)
else:
    raise Exception("Failed to start server.")
```

File: mlflow/sagemaker/cli.py [hunk 12]
```diff
@@ -358,22 +358,26 @@ def build_and_push_container(build, push, container, env_manager, mlflow_home):
         click.echo("skipping both build and push, have nothing to do!")
     if build:
         sagemaker_image_entrypoint = (
-            'ENTRYPOINT ["python", "-c", "import sys; from mlflow.models import container as C; '
-            f"C._init(sys.argv[1], '{env_manager}')\"]"
+            "import sys; from mlflow.models import container as C; "
+            f"C._init(sys.argv[1], '{env_manager}')"
```
Comment (harupy) [inline mlflow/sagemaker/cli.py:362]: Unrelated to this PR, creating a module like below and running that using `python -m mlflow.foo.bar` might be cleaner and harder to make mistakes:

```
import sys
from mlflow.models import container

container._init(sys.argv[1], sys.argv[2])
```

File: tests/pyfunc/docker/conftest.py (hunk 4 above)
Comment (harupy) [reply tests/pyfunc/docker/conftest.py:29]: Can we catch `docker.errors.ImageNotFound` instead of `Exception`?

Comment (harupy) [reply tests/pyfunc/docker/conftest.py:29]: `docker_client.containers.list` seems to return an empty list when `filters` matches nothing. Can we pull this out from try-catch?

File: tests/pyfunc/docker/test_docker_flavors.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
+"""
```
Comment (harupy) [inline tests/pyfunc/docker/test_docker_flavors.py:1]: Is this something we should run nightly?

File: tests/pyfunc/docker/test_docker_flavors.py [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+        mlflow_home=MLFLOW_ROOT,  # Required to prevent installing dev version of MLflow from PyPI
+    )
+
+    # Run a container
+    port = get_safe_port()
+    docker_client.containers.run(
+        image=TEST_IMAGE_NAME,
+        ports={8080: port},
+        detach=True,
+    )
```
Comment (harupy) [inline tests/pyfunc/docker/test_docker_flavors.py:104]: Can we kill this container when it's no longer needed?

File: tests/pyfunc/docker/conftest.py (hunk 4 above)
Comment (B-Step62) [reply tests/pyfunc/docker/conftest.py:29]: yeah scoping down try-catch makes sense!

File: mlflow/sagemaker/cli.py (hunk 12 above)
Comment (B-Step62) [reply mlflow/sagemaker/cli.py:362]: Yup we may want to do that when the script becomes more complex. Let me keep it as it is in the scope of this PR.

File: tests/pyfunc/docker/test_docker_flavors.py (hunk 13 above)
Comment (B-Step62) [reply tests/pyfunc/docker/test_docker_flavors.py:1]: Yes! As suggested by Ben, I created a follow-up JIRA for this sprint to run this periodically like cross version tests.

File: tests/pyfunc/docker/test_docker_flavors.py (hunk 14 above)
Comment (B-Step62) [reply tests/pyfunc/docker/test_docker_flavors.py:104]: Could you elaborate a bit? They are cleaned up in fixture at the end of each test case now.

**Review by harupy — APPROVED** [review]
Left a few comments, otherwise LGTM!

File: tests/pyfunc/docker/test_docker_flavors.py (hunk 14 above)
Comment (harupy) [reply tests/pyfunc/docker/test_docker_flavors.py:104]: Ah I see, `clean_up_docker_image` is in `tests/pyfunc/docker/conftest.py`.

**Comment by B-Step62** [comment]
Using virtual Windows machine for testing was quite a bit of effort - permission setting, installing tools, etc. I ended up testing with my personal Windows laptop:p

The basic `test_docker.py` all passed.
capture`

For flavors, mostly passed but failed with a few
1. `spark`: Failed with spark installation indeed, not relevant to this change (also we don't introduce any change for Java flavors).
2. `tensorflow/keras/transformers`: Basically those depends on Tensorflow. The reason is that the model is logged on Windows so with `tensorflow-intel` as requirement, while it's not available in the container based on Ubuntu. Should be unrelated to the change itself.

So overall I think this change shouldn't introduce new surprise for Windows users:)

**Comment by marxxxx** [comment]
Is this optimization already available with mlflow-2.15.1? 

I just trained a simple sample model with this tutorial (https://github.com/Azure/azureml-examples/blob/main/tutorials/azureml-in-a-day/azureml-in-a-day.ipynb). The model itself is ~120kb in size but the docker image is still 3.8 GB
image

I also trained an object detection model which was ~12 GB in size. As we are planning to do interference on edge devices this might be way to large for us.
